src/checkpointing.jl

```diff
@@ -0,0 +1,29 @@
+"""
+    checkpoint(f, args::Tuple)
+
+Treat `f` as a primitive, whose gradient is computed by running and forwards- and reverse-
+pass of ∇ inside the outer pass. This can have the effect of significantly lowering the
+peak memory usage of a gradient computation at the expense of additional computation.
+
+HEALTH WARNING: If you close over differentiable objects in `f`, you will get incorrect
+results. It is therefore very much recommended to verify your gradients using finite
+differencing whenever code is checkpointed.
+"""
+checkpoint(f, args::Tuple) = f(args...)
+
+@explicit_intercepts checkpoint Tuple{Any, Any} [false, true]
+
+∇(::typeof(checkpoint), ::Type{Arg{1}}, p, y, ȳ, f, args::Tuple) = nothing
```

Comment:
Shouldn't be necessary to define a method for `Arg{1}` since you've already said that we aren't intercepting the first argument.